### PR TITLE
qa/cephadm: Add RGW

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Fixes: https://tracker.ceph.com/issues/43911

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by mgfritch** [comment]
jenkins test make check

**Comment by sebastian-philipp** [comment]
> It appears as if we don't have any smoke test for nfs?

yep. I think nfs is not tested at all in teuthology right now. That's until @varshar16 is ready.

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com/swagner-2020-05-15_11:46:15-rados-wip-swagner3-testing-2020-05-15-1059-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
```
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/contextutil.py", line 32, in nested
    vars.append(enter())
  File "/usr/lib/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-swagner3-testing-2020-05-15-1059/qa/tasks/cephadm.py", line 713, in ceph_rgw
    (realm, zone) = realmzone.split('.', 1)
ValueError: not enough values to unpack (expected 2, got 1)
```

**Comment by sebastian-philipp** [comment]
something is fishy:

```
/home/ubuntu/cephtest/cephadm --image quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid f4a812f0-96f5-11ea-a069-001a4aab830c -- bash -c 'ceph orch ps'
NAME                                                      HOST       STATUS          REFRESHED  AGE   VERSION                   IMAGE NAME                                                     IMAGE ID      CONTAINER ID
alertmanager.a                                            smithi077  running (69s)   60s ago    96s   0.20.0                    prom/alertmanager                                              0881eb8f169f  d8a679638650
grafana.a                                                 smithi061  running (80s)   56s ago    80s   6.6.2                     ceph/ceph-grafana:latest                                       87a51ecf0b1c  75743178e5b2
mgr.x                                                     smithi061  running (5m)    56s ago    5m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  1f861ed0c85e
mgr.y                                                     smithi077  running (7m)    60s ago    7m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  9dc339b920dd
mon.a                                                     smithi077  running (7m)    60s ago    7m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  e6f28c6dfc64
mon.b                                                     smithi061  running (5m)    56s ago    5m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  13f981993314
mon.c                                                     smithi077  running (6m)    60s ago    6m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  2a1c17fdb2c1
node-exporter.a                                           smithi077  running (108s)  60s ago    110s  0.18.1                    prom/node-exporter                                             e5a616e4b9cf  06da224d2223
node-exporter.b                                           smithi061  running (109s)  56s ago    111s  0.18.1                    prom/node-exporter                                             e5a616e4b9cf  f33752836fa3
osd.0                                                     smithi077  running (5m)    60s ago    5m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  0be1d0e0d9bf
osd.1                                                     smithi077  running (4m)    60s ago    4m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  91f629c331ac
osd.2                                                     smithi077  running (4m)    60s ago    4m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  eecd29df2f30
osd.3                                                     smithi077  running (3m)    60s ago    3m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  7e9d42f9b570
osd.4                                                     smithi061  running (3m)    56s ago    3m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  a961fa0b7895
osd.5                                                     smithi061  running (3m)    56s ago    3m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  3bb56d6ceab6
osd.6                                                     smithi061  running (2m)    56s ago    2m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  bddc9c16af15
osd.7                                                     smithi061  running (2m)    56s ago    2m    16.0.0-1580-g0b73b0676ce  quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58  5a635ebde391
prometheus.a                                              smithi061  running (62s)   56s ago    111s  2.18.1                    prom/prometheus:latest                                         de242295e225  7f729cf7c78e
rgw.realm.zone.1;smithi077=realm.zone.a.smithi061.dwpprr  smithi061  unknown         56s ago    2m    <unknown>                 quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  <unknown>     <unknown>
rgw.realm.zone.1;smithi077=realm.zone.a.smithi077.majyrb  smithi077  unknown         60s ago    2m    <unknown>                 quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  <unknown>     <unknown>
/home/ubuntu/cephtest/cephadm --image quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid f4a812f0-96f5-11ea-a069-001a4aab830c -- bash -c 'ceph orch ls'
NAME                                     RUNNING  REFRESHED  AGE   PLACEMENT                        IMAGE NAME                                                     IMAGE ID
alertmanager                                 1/1  64s ago    2m    count:1 smithi077=a              prom/alertmanager                                              0881eb8f169f
grafana                                      1/1  60s ago    117s  count:1 smithi061=a              ceph/ceph-grafana:latest                                       87a51ecf0b1c
mgr                                          2/2  64s ago    5m    count:2 smithi061=x              quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58
mon.a                                        3/0  64s ago    -     <unmanaged>                      quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58
node-exporter                                2/2  64s ago    2m    count:2 smithi077=a,smithi061=b  prom/node-exporter                                             e5a616e4b9cf
osd.0                                        8/0  64s ago    -     <unmanaged>                      quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  70faa3111d58
prometheus                                   1/1  60s ago    2m    count:1 smithi061=a              prom/prometheus:latest                                         de242295e225
rgw.realm.zone.1;smithi077=realm.zone.a      0/2  64s ago    2m    count:2                          quay.io/ceph-ci/ceph:0b73b0676ce593ad325d24ec36d8085b7e2f644d  <unknown>
```

And: https://tracker.ceph.com/issues/44168 again:
```
2020-05-15T21:58:57.654 INFO:tasks.cephadm:Teardown complete
2020-05-15T21:58:57.655 ERROR:teuthology.run_tasks:Manager failed: cephadm
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 168, in run_tasks
    suppress = manager.__exit__(*exc_info)
  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
    next(self.gen)
  File "/home/teuthworker/src/github.com_sebastian-philipp_ceph_qa-cephadm-add-rgw-and-iscsi/qa/tasks/cephadm.py", line 1112, in task
    log.info('Teardown begin')
  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
    next(self.gen)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/contextutil.py", line 56, in nested
    reraise(*exc)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/lib/python3.6/site-packages/six.py", line 702, in reraise
    raise value.with_traceback(tb)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/contextutil.py", line 48, in nested
    if exit(*exc):
  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
    next(self.gen)
  File "/home/teuthworker/src/github.com_sebastian-philipp_ceph_qa-cephadm-add-rgw-and-iscsi/qa/tasks/cephadm.py", line 454, in ceph_bootstrap
    ctx.daemons.get_daemon(type_, id_, cluster).stop()
AttributeError: 'NoneType' object has no attribute 'stop'
```

**Comment by sebastian-philipp** [comment]
ttp://pulpito.ceph.com:80/swagner-2020-05-18_10:04:49-rados-wip-swagner3-testing-2020-05-15-1059-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com:80/swagner-2020-05-18_10:51:25-rados-wip-swagner3-testing-2020-05-15-1059-distro-basic-smithi/

File: qa/tasks/cephadm.py [hunk 1]
```diff
@@ -451,7 +451,11 @@ def ceph_bootstrap(ctx, config):
         # so, stop them individually
         for role in ctx.daemons.resolve_role_list(None, CEPH_ROLE_TYPES):
             cluster, type_, id_ = teuthology.split_role(role)
-            ctx.daemons.get_daemon(type_, id_, cluster).stop()
+            try:
+                ctx.daemons.get_daemon(type_, id_, cluster).stop()
+            except AttributeError:
+                log.error(f'Failed to stop {(cluster, type_, id_)} "{role}" {ctx.daemons.daemons.keys()}')
```
Comment (sebastian-philipp) [inline qa/tasks/cephadm.py:457]: This prints:

```
Failed to stop ('rgw', 'realm', 'zone.a') "rgw.realm.zone.a" dict_keys(['ceph.mon', 'ceph.mgr', 'ceph.osd', 'ceph.rgw', 'ceph.prometheus', 'ceph.node-exporter', 'ceph.alertmanager', 'ceph.grafana'])
```

which means we have `cluster == 'rgw'` which breaks this routine. 

@jdurgin or  @liewegas  any ideas? Is this a bug in `split_role`?

Comment (jdurgin) [reply qa/tasks/cephadm.py:457]: I think the simplest way to make this work is to change the resolve_role_list() call to use cluster_aware=true, like the rest of the callers in cephadm.py and ceph.py.

Then you'll get 'client.rgw.realm.zone.a', which split_role() will handle correctly.

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com:80/swagner-2020-05-19_10:43:44-rados-wip-swagner3-testing-2020-05-15-1059-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com:80/swagner-2020-05-19_11:57:30-rados-wip-swagner2-testing-2020-05-18-1100-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
we're getting somewhere:

```
"Module 'cephadm' has failed: DaemonDescription: Cannot calculate service_id: daemon_id='realm.zone.a' hostname='smithi028'","count":1},"muted":false}},"mutes":[]}
```

File: src/pybind/mgr/orchestrator/_interface.py [hunk 2]
```diff
@@ -1314,6 +1314,11 @@ def _match():
                 if m:
                     return m.group(1)
 
+            if self.daemon_type == 'rgw':
+                v = self.daemon_id.split('.')
+                if len(v) in [3, 4]:
+                    return '.'.join(v[0:2])
+
```
Comment (mgfritch) [inline src/pybind/mgr/orchestrator/_interface.py:1321]: :+1:   cherry-picked this into #35142, but can rebase the change out if you'd like to merge this first

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator/_interface.py:1321]: I think we can merge yours

**Comment by sebastian-philipp** [comment]
```2020-05-20T13:07:29.580 INFO:ceph.rgw.realm.zone.a.smithi203.stdout:May 20 13:07:29 smithi203 bash[18677]: debug 2020-05-20T13:07:29.379+0000 7f92b4028240 -1 Cannot find zone id= (name=zone)
2020-05-20T13:07:29.580 INFO:ceph.rgw.realm.zone.a.smithi203.stdout:May 20 13:07:29 smithi203 bash[18677]: debug 2020-05-20T13:07:29.380+0000 7f92b4028240 -1 Couldn't init storage provider (RADOS)
```

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com/swagner-2020-05-29_10:22:55-rados:cephadm-wip-swagner-testing-2020-05-27-1352-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com:80/swagner-2020-05-29_11:01:42-rados:cephadm-wip-swagner-testing-2020-05-27-1352-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
success: http://pulpito.ceph.com/swagner-2020-05-29_11:01:42-rados:cephadm-wip-swagner-testing-2020-05-27-1352-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
jenkins run make check

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com/swagner-2020-06-02_12:46:10-rados:cephadm-wip-swagner3-testing-2020-06-02-1145-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Pull request merged**